### PR TITLE
fix filter for 4.2

### PR DESCRIPTION
**Pull request by zengmin-wish**
for https://jira.wish.site/browse/CORE-1269

### CODE REVIEW

File: mongoengine/document.py [hunk 1]
```diff
@@ -167,7 +167,7 @@ def save(self, safe=True, force_insert=None, validate=True):
                     if self._get_write_decider():
                         proxy_client.instance().update(
                             self.__class__,
-                            {"_id" : doc["_id"]},
+                            self.__class__.get_upsert_filter(doc),
```
Comment (yiguo-wish) [inline mongoengine/document.py:170]: I think we can use `cls.get_upsert_filter(doc)`

File: tests/document.py [hunk 2]
```diff
@@ -1010,6 +1010,44 @@ class NonIdShardedCollection(Document):
...
+                          { "_id": "abcd" })
+	    self.assertEquals(IdShardedCollection.get_upsert_filter({"_id":"abcd", "foo":"bar"}),
+                          { "_id": "abcd" })
+	    self.assertEquals(IdShardedCollection2.get_upsert_filter({"_id":"abcd", "n":"bar"}),
+                          { "_id": "abcd", "n":"bar" })
+	    self.assertEquals(IdShardedCollection3.get_upsert_filter({"_id":"abcd", "n":"bar"}),
+                          { "_id": "abcd" })
+	    self.assertEquals(IdShardedCollection2.get_upsert_filter({"_id":"abcd"}),
+                          { "_id": "abcd" })
+        self.assertEquals(NonIdShardedCollection.get_upsert_filter({"_id":"abcd", "n": "foo"}),
```
[the remark this answers is not in the document]
Comment (zengmin-wish) [reply tests/document.py:1046]: fixed

File: mongoengine/document.py (hunk 1 above)
Comment (zengmin-wish) [reply mongoengine/document.py:170]: no, cls is not defined here.

Comment (yiguo-wish) [reply mongoengine/document.py:170]: ah, the caller is instance method. actually class method could be called with instance obj. so probably self.get_upsert_filter(doc) works as well

https://docs.python.org/2/library/functions.html#classmethod
```A class method can be called either on the class (such as C.f()) or on an instance (such as C().f()).```

https://realpython.com/instance-class-and-static-methods-demystified/

**Review by yiguo-wish — APPROVED** [review]

File: mongoengine/document.py [hunk 3]
```diff
@@ -188,6 +188,19 @@ def save(self, safe=True, force_insert=None, validate=True):
         id_field = self._meta['id_field']
         self[id_field] = self._fields[id_field].to_python(object_id)
 
+    # For Mongo 4.2 and above, the filter must contain shard key if upsert is true.
+    @classmethod
+    def get_upsert_filter(cls, doc):
+        filter = {"_id" : doc["_id"]}
+        shard_keys = cls._meta.get('shard_key', "")
+        shard_fields = [s.split(':')[0] for s in shard_keys.split(',')]
```
Comment (tsitu-wish) [inline mongoengine/document.py:196]: @zengmin-wish  I think this is causing issues when shard_key is set to `False`. We are seeing login failuires on merchant side.

```
      File "/home/app/clroot/cl/utils/tornadoutil/abstract.py", line 556, in call
        method.data = yield gen.Task(method.run)
      File "/home/app/virtualenv/lib/python2.7/site-packages/tornado/gen.py", line 1055, in run
        value = future.result()
      File "/home/app/virtualenv/lib/python2.7/site-packages/tornado/concurrent.py", line 238, in result
        raise_exc_info(self._exc_info)
      File "/home/app/virtualenv/lib/python2.7/site-packages/tornado/gen.py", line 622, in Task
        func(*args, callback=_argument_adapter(set_result), **kwargs)
      File "/home/app/clroot/sweeper/merchant_dashboard/api/authentication/login.py", line 136, in run
        merchant_user_instance=merchant_user
      File "/home/app/clroot/sweeper/merchant_dashboard/lib/authentication/login.py", line 259, in login_user_from_web_handler
        login_response = login_user(username, password, ip_address, **kwargs)
      File "/home/app/clroot/sweeper/merchant_dashboard/lib/authentication/login.py", line 707, in login_user
        oauth_credential_id=oauth_credential_id,
      File "/home/app/clroot/sweeper/merchant_dashboard/model/session.py", line 178, in create_session_for_user
        session.save()
      File "/home/app/virtualenv/lib/python2.7/site-packages/mongoengine/document.py", line 170, in save
        self.__class__.get_upsert_filter(doc),
      File "/home/app/virtualenv/lib/python2.7/site-packages/mongoengine/document.py", line 196, in get_upsert_filter
        shard_fields = [s.split(':')[0] for s in shard_keys.split(',')]
    AttributeError: 'bool' object has no attribute 'split'
```

**Pull request merged**